barquesubmit/client.go

```diff
@@ -0,0 +1,203 @@
+package barquesubmit
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/evergreen-ci/gimlet"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/curator/repobuilder"
+	"github.com/pkg/errors"
+)
+
+const (
+	barqueAPIKeyHeader  = "Api-Key"
+	barqueAPIUserHeader = "Api-User"
+)
+
+type Client struct {
+	baseURL  string
+	username string
+	apiKey   string
+}
+
+func New(baseURL string) (*Client, error) {
+	if !strings.HasPrefix(baseURL, "http") {
+		return nil, errors.New("malformed url")
+	}
+
+	if !strings.HasSuffix(baseURL, "/") {
+		baseURL += "/"
+	}
+
+	if !strings.HasSuffix(baseURL, "/rest/v1/") {
+		baseURL += "rest/v1/"
+	}
+
+	return &Client{
+		baseURL: baseURL,
+	}, nil
+}
+
+func (c *Client) getURL(p string) string {
+	if strings.HasPrefix(p, c.baseURL) {
+		return p
+	}
+
+	return path.Join(c.baseURL, p)
+}
+
+func (c *Client) makeRequest(ctx context.Context, url, method string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest(method, c.getURL(url), body)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem with request")
+	}
+	req = req.WithContext(ctx)
+
+	if c.apiKey == "" {
+		return req, nil
+	}
+
+	if c.username != "" {
+		req.Header[barqueAPIUserHeader] = []string{c.username}
+	}
+	if c.apiKey != "" {
+		req.Header[barqueAPIKeyHeader] = []string{c.apiKey}
+	}
+
+	return req, nil
+}
+
+type userCredentials struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+type userAPIKeyResponse struct {
+	Username string `json:"username"`
+	Key      string `json:"key"`
+}
+
+func (c *Client) Login(ctx context.Context, username, password string) error {
+	client := utility.GetDefaultHTTPRetryableClient()
+	defer utility.PutHTTPClient(client)
+
+	payload, err := json.Marshal(&userCredentials{Username: username, Password: password})
+	if err != nil {
+		return errors.Wrap(err, "problem marshaling login payload")
+	}
+
+	req, err := c.makeRequest(ctx, "admin/login", http.MethodGet, bytes.NewBuffer(payload))
+	if err != nil {
+		return errors.Wrap(err, "problem building login request")
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "problem making login request")
+	}
+	defer resp.Body.Close()
```

Comment:
I don't think this is necessary since all code paths below this end up calling gimlet.GetJSON(), which closes the response body.

Reply:
what if we return before calling getJSON? I think we've had bugs with this in the past.